build_runner/lib/src/package_graph/package_graph.dart

```diff
@@ -117,6 +119,40 @@ class PackageGraph {
   /// Shorthand to get a package by name.
   PackageNode operator [](String packageName) => allPackages[packageName];
 
+  /// Finds all packages which depend on [packageName] in postorder by
+  /// dependencies.
+  ///
+  /// See [orderedPackages] for ordering guarantees. The node for [packageName]
+  /// will not be included in the result.
+  Iterable<PackageNode> dependentsOf(String packageName) {
+    if (!allPackages.containsKey(packageName)) return const [];
+    var node = allPackages[packageName];
+    return orderedPackages.where((n) => n.dependencies.contains(node));
+  }
+
+  /// All of the packages in postorder by dependencies.
+  ///
+  /// Depedenencies of a package will come before the package in the result. If
+  /// there is a package cycle the relative position of packages within the
+  /// cycle is non-deterministic, ecept that the root package will always come
+  /// last. For any two packages for which neither is a transitive dependency of
+  /// the other the relative position of the packages within the cycle is
+  /// non-deterministic.
+  Iterable<PackageNode> get orderedPackages sync* {
+    var seen = new Set<PackageNode>();
+    var queue = new Queue<PackageNode>();
+    queue.addLast(root);
+    while (queue.isNotEmpty) {
+      var next = queue.last;
+      if (next.dependencies.isEmpty || seen.contains(next)) {
+        yield queue.removeLast();
```

Comment:
I think this logic will result in duplicate packages if you have a shared dependency between two packages (for instance if `a` is the root and depends on `b`, and `c`, and both `b` and `c` depend on `d`, then `d` will get yielded multiple times).

To me the recursive logic is easier to follow than this iterative version (especially with the queue and sync* etc), and we shouldn't ever have so many packages that a recursive algorithm could run into issues. That is just my preference though.

```dart
Iterable<PackageNode> get orderedPackages {
  var packages = <PackageNode>[];
  var seen = new Set<PackageNode>();

  void crawlDeps(PackageNode package) {
    seen.add(package);
    for (var dep in package.dependencies) {
      if (seen.contains(dep)) continue;
      crawlDeps(dep);
    }
    packages.add(package);
  }
  crawlDeps(root);
  
  return packages;
}
```

Reply:
yeah, I can fix the bug with duplicates easily but it's not as clear to me how to fix the bug with the root package in a cycle. I switched back to the recursive version and we can hope that we never see deep enough package graphs where it's an issue.

I did keep `sync*` because IMO it's more clear than the closure and `yield*` works nicely with the recursive call.